### PR TITLE
Add proxy and mirror configuration for etcd machines

### PR DESCRIPTION
**Pull request by mrajashree**

### CODE REVIEW

File: pkg/userdata/bottlerocket/etcd_init.go [hunk 1]
```diff
@@ -18,10 +17,10 @@ runcmd: "{{ .EtcdadmInitCommand }}"
 `
 
 // NewInitEtcdPlane returns the user data string to be used on a etcd instance.
-func NewInitEtcdPlane(input *userdata.EtcdPlaneInput, config bootstrapv1alpha3.BottlerocketConfig, log logr.Logger) ([]byte, error) {
+func NewInitEtcdPlane(input *userdata.EtcdPlaneInput, config userdata.Config, log logr.Logger) ([]byte, error) {
```
Comment (g-gaston) [inline pkg/userdata/bottlerocket/etcd_init.go]: At this point, why not just pass the whole `bootstrapv1alpha3.EtcdadmConfig`?
If we want a separate struct, then I would vote to make it independent from the api structs. That would have the added benefit of being able to extract this code as a separate module that can be reused in difference controllers

File: pkg/userdata/bottlerocket/node_userdata.go [hunk 2]
```diff
@@ -36,20 +37,42 @@ superpowered = true
...
 {{- end -}}
+`
+
+	networkInitTemplate = `{{ define "networkInitSettings" -}}
+[settings.network]
+https-proxy = "{{.HTTPSProxyEndpoint}}"
+no-proxy = "{{.NoProxyEndpoints}}"
+{{- end -}}
+`
+	registryMirrorTemplate = `{{ define "registryMirrorSettings" -}}
```
Comment (g-gaston) [inline pkg/userdata/bottlerocket/node_userdata.go:47]: where is this one and `registryMirrorCACertSettings` being used?

Comment (mrajashree) [reply pkg/userdata/bottlerocket/node_userdata.go:47]: my bad, forgot to push the update earlier

**Review by g-gaston — APPROVED** [review]
🚀

File: pkg/userdata/bottlerocket/node_userdata.go [hunk 3]
```diff
@@ -75,15 +109,31 @@ func generateBottlerocketNodeUserData(bootstrapContainerUserData []byte, users [
...
 		AdminContainerUserData:     b64AdminContainerUserData,
-		BootstrapContainerSource:   config.BootstrapImage,
-		PauseContainerSource:       config.PauseImage,
+		BootstrapContainerSource:   config.BottlerocketConfig.BootstrapImage,
+		PauseContainerSource:       config.BottlerocketConfig.PauseImage,
+	}
+
+	if config.Proxy != nil {
+		bottlerocketInput.HTTPSProxyEndpoint = config.Proxy.HTTPSProxy
+		if len(config.Proxy.NoProxy) > 0 {
```
Comment (g-gaston) [inline pkg/userdata/bottlerocket/node_userdata.go]: nit: do you need this? even if the array is empty or nil, the for loop should already take care of that

File: pkg/userdata/cloudinit/cloudinit.go [hunk 4]
```diff
@@ -81,3 +103,62 @@ func buildEtcdadmArgs(config bootstrapv1alpha3.CloudInitConfig) userdata.Etcdadm
...
+	}
+	tmpl := template.New("proxy").Funcs(template.FuncMap{"stringsJoin": strings.Join})
+	t, err := tmpl.Parse(proxyConf)
+	if err != nil {
+		return fmt.Errorf("failed to parse proxy template: %v", err)
+	}
+
+	var out bytes.Buffer
+	err = t.Execute(&out, proxy)
+	if err != nil {
```
Comment (g-gaston) [inline pkg/userdata/cloudinit/cloudinit.go]: nit:
```suggestion
	if err = t.Execute(&out, proxy); err != nil {
```

File: pkg/userdata/cloudinit/cloudinit.go [hunk 5]
```diff
@@ -81,3 +103,62 @@ func buildEtcdadmArgs(config bootstrapv1alpha3.CloudInitConfig) userdata.Etcdadm
...
+	}
+	tmpl := template.New("registryMirror")
+	t, err := tmpl.Parse(registryMirrorConf)
+	if err != nil {
+		return fmt.Errorf("failed to parse registryMirror template: %v", err)
+	}
+
+	var out bytes.Buffer
+	err = t.Execute(&out, registryMirror)
+	if err != nil {
```
Comment (g-gaston) [inline pkg/userdata/cloudinit/cloudinit.go]: same here about a one liner

File: pkg/userdata/cloudinit/cloudinit.go [hunk 6]
```diff
@@ -81,3 +103,62 @@ func buildEtcdadmArgs(config bootstrapv1alpha3.CloudInitConfig) userdata.Etcdadm
...
+		{
+			Content: registryMirror.CACert,
+			Owner:   "root:root",
+			Path:    fmt.Sprintf("/etc/containerd/certs.d/%s/ca.crt", registryMirror.Endpoint),
+		}, {
+			Content: out.String(),
+			Owner:   "root:root",
+			Path:    "/etc/containerd/config_append.toml",
+		},
+	}...)
```
Comment (g-gaston) [inline pkg/userdata/cloudinit/cloudinit.go]: why do we need to create a slice to then serialize it? what about:
```suggestion
	input.AdditionalFiles = append(input.AdditionalFiles,
		capbk.File{
			Content: registryMirror.CACert,
			Owner:   "root:root",
			Path:    fmt.Sprintf("/etc/containerd/certs.d/%s/ca.crt", registryMirror.Endpoint),
		},
		capbk.File{
			Content: out.String(),
			Owner:   "root:root",
			Path:    "/etc/containerd/config_append.toml",
		},
	)
```

File: pkg/userdata/cloudinit/cloudinit.go [hunk 7]
```diff
@@ -81,3 +103,62 @@ func buildEtcdadmArgs(config bootstrapv1alpha3.CloudInitConfig) userdata.Etcdadm
...
+
+	input.PreEtcdadmCommands = append(input.PreEtcdadmCommands, `cat /etc/containerd/config_append.toml >> /etc/containerd/config.toml`)
+	input.PreEtcdadmCommands = append(input.PreEtcdadmCommands, containerdRestart...)
```
Comment (g-gaston) [inline pkg/userdata/cloudinit/cloudinit.go:161]: nit: you could combine these two appends

**Review by g-gaston — APPROVED** [review]
I like it, minor comments

File: pkg/userdata/bottlerocket/node_userdata.go (hunk 3 above)
Comment (mrajashree) [reply pkg/userdata/bottlerocket/node_userdata.go]: yup don't need this, I think it got left after some trial-error to get this working

File: pkg/userdata/cloudinit/cloudinit.go (hunk 6 above)
Comment (mrajashree) [reply pkg/userdata/cloudinit/cloudinit.go]: I changed it but aren't they both effectively the same?

File: pkg/userdata/cloudinit/cloudinit.go (hunk 7 above)
Comment (mrajashree) [reply pkg/userdata/cloudinit/cloudinit.go:161]: I had tried that but it doesn't allow append where the third parameter being variadic

**Pull request merged**